codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -0,0 +1,280 @@
+            Http2DataFrame dataFrame = (Http2DataFrame) frame;
+            http2Handler.encoder().writeData(http2HandlerCtx, streamId, dataFrame.content().retain(),
+                                             dataFrame.padding(), dataFrame.isEndStream(), promise);
+        } else if (frame instanceof Http2HeadersFrame) {
+            Http2HeadersFrame headerFrame = (Http2HeadersFrame) frame;
+            http2Handler.encoder().writeHeaders(
+                    http2HandlerCtx, streamId, headerFrame.headers(), headerFrame.padding(), headerFrame.isEndStream(),
+                    promise);
+        } else if (frame instanceof Http2ResetFrame) {
+            Http2ResetFrame rstFrame = (Http2ResetFrame) frame;
+            http2Handler.resetStream(http2HandlerCtx, streamId, rstFrame.errorCode(), promise);
+        } else {
+            throw new UnsupportedMessageTypeException(frame);
+        }
+    }
+
+    private final class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(stream.id()));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            ctx.fireUserEventTriggered(new Http2StreamClosedEvent(stream.id()));
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData);
+            ctx.fireChannelRead(goAway);
+        }
+    }
+
+    private final class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        InternalHttp2ConnectionHandler(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
+                                       Http2Settings initialSettings) {
+            super(decoder, encoder, initialSettings);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                                     Http2Exception.StreamException http2Ex) {
+            try {
+                Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
+                if (stream == null) {
+                    return;
+                }
+                ctx.fireExceptionCaught(http2Ex);
+            } finally {
+                super.onStreamError(ctx, cause, http2Ex);
+            }
+        }
+    }
+
+    private final class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode) {
+            Http2ResetFrame rstFrame = new DefaultHttp2ResetFrame(errorCode);
+            rstFrame.setStreamId(streamId);
```

Comment:
consider initializing the stream id in the constructor where possible, and only using the `setStreamId()` method where necessary ... maybe a bit more constructor code but it may help clarify the expected stream lifetime and the need for the setter method.